batch_job/src/batch_job_amazon.c

```diff
@@ -85,20 +108,136 @@ the desired amount of cores, memory, and disk.  Return
 the name of the instance, if one exists, otherwise null.
 */
 
-
 static const char * aws_instance_select( int cores, int memory, int disk )
 {
 	struct aws_instance_type *i;
 
 	for(i=aws_instance_table;i->name;i++) {
 		if(cores<=i->cores && memory<=i->memory) {
-			debug(D_BATCH,"job requiring CORES=%d MEMORY=%d matches instance type %s",cores,memory,i->name);
+			debug(D_BATCH,"job requiring CORES=%d MEMORY=%d matches instance type %s\n",cores,memory,i->name);
 			return i->name;
 		}
 	}
 	return 0;
 }
 
+/*
+Lookup aws_instance_type given instance type name
+*/
+
+struct aws_instance_type * aws_instance_lookup(const char *instance_type)
+{
+	struct aws_instance_type *i;
+	for(i=aws_instance_table;i->name;i++) {
+		if(strcmp(instance_type, i->name) == 0) {
+			return i;
+		}
+	}
+	debug(D_BATCH,"instance type %s not found in instance type table\n", instance_type);
+	return 0;
+}
+
+/*
+Compare if instance of type 1 does fit in instance of type 2.
+On fitting (cores and memory of type 1 instance <= type 2)
+return 1, otherwise return null
+*/
+
+int compare_instance_type(const char* type_1, const char* type_2){
+	if(strcmp(type_1,type_2) == 0){
+		return 1;
+	} else {
+		struct aws_instance_type *instance_1 = aws_instance_lookup(type_1);
+		struct aws_instance_type *instance_2 = aws_instance_lookup(type_2);
+		if (instance_1->cores<=instance_2->cores && instance_1->memory<=instance_2->memory){
+			return 1;
+		}
+	}
+	return 0;
+}
+
+struct idle_instance* record_idle_instance(const char *instance_id, const char *instance_type){
+	struct idle_instance* i = malloc(sizeof(*i));
+	i->instance_id = instance_id;
+	i->instance_type = instance_type;
+	i->next = NULL;
+	return i;
+}
+
+/*
+Push an idle instance to the end of the list
+return 0 on error and 1 on success
+*/
+
```

Comment:
Save yourself some code here by using the linked list implementation in `dttools/src/list.h`